### ISSUE TITLE
$delete does not work unless namespace is set

### ISSUE THREAD

**Issue by bcbrockway** [kubernetes-sigs/kustomize#1343]
I have a base containing a deployment that I want to patch in an overlay:

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysqld-exporter
spec:
  template:
    spec:
      containers:
      - name: mysqld-exporter
        image: prom/mysqld-exporter:v0.11.0
        command:
        - /bin/mysqld_exporter
        args:
        - --web.listen-address=:9104
        - --log.level=info
        - --log.format=logger:stderr?json=true
        livenessProbe:
          httpGet:
            path: /
            port: 9104
          initialDelaySeconds: 30
        readinessProbe:
          httpGet:
            path: /
            port: 9104
        env:
        - name: DATA_SOURCE_NAME
          value: $(DB_USER):$(DB_PASSWORD)@(127.0.0.1:3306)/
        ports:
        - name: metrics
          containerPort: 9104
        resources:
          requests:
            cpu: 30m
            memory: 50Mi
          limits:
            cpu: 100m
            memory: 200Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 65534
      - name: cloud-sql-proxy
        image: gcr.io/cloudsql-docker/gce-proxy
        imagePullPolicy: IfNotPresent
        command:
        - /cloud_sql_proxy
        args:
        - -instances=$(MASTER_INSTANCE)=tcp:127.0.0.1:3306
        - -credential_file=/secrets/cloudsql/google_credentials
        - -dir=/cloudsql
        ports:
        - name: mysql
          containerPort: 3306
        resources:
          requests:
            cpu: 30m
            memory: 50Mi
          limits:
            cpu: 100m
            memory: 200Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 65534
        volumeMounts:
        - name: cloudsql-settings
          mountPath: /secrets/cloudsql
          readOnly: true

```

In my overlay I have one patch that removes the cloud-sql-proxy container:

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysqld-exporter
  namespace: mynamespace
spec:
  template:
    spec:
      containers:
      - $patch: delete
        name: cloud-sql-proxy

```

and one that adds some variables to the mysqld-exporter container:

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysqld-exporter
spec:
  template:
    spec:
      containers:
      - name: mysqld-exporter
        env:
        - name: DATA_SOURCE_NAME
          value: user:password@(host:3306)/
```

This works, but if you take the namespace out of the $delete patch it finds the resource when you do a `kustomize build .` (as evidenced by the absence of a `failed to find an object with apps_v1_Deployment|mysqld-exporter to apply the patch` error), but doesn't delete the container. If you add the namespace back in it _does_ delete it.

The other patch adds the variables whether or not the namespace is provided in the patch. This had us scratching our heads for a while trying to figure out why the $delete patch wouldn't work. If the add-vars patch didn't need a namespace then I don't think the $delete patch should either.

**Comment by richardmarshall** [comment]
Can you please provide the `kustomization.yaml` that you have for the base and overlay?

**Comment by bcbrockway** [comment]
Hi Richard, yeah there are actually a few levels to this. There's an initial `mysqld-exporter` base - we use this as a template from which to derive distinct instances:

```yaml
kind: Kustomization
apiVersion: kustomize.config.k8s.io/v1beta1

commonLabels:
  app.kubernetes.io/component: prometheus-exporter

configurations:
- commonlabels.yaml

resources:
- resources/mysqld-exporter-deployment.yaml
- resources/mysqld-exporter-service.yaml
- resources/mysqld-exporter-serviceMonitor.yaml
```

This gets inherited by an instance-specific overlay containing common config for all environments. This is where we set the namespace:

```yaml
---
kind: Kustomization
apiVersion: kustomize.config.k8s.io/v1beta1

bases:
- ../../../templates/mysqld-exporter

commonLabels:
  app.kubernetes.io/managed-by: pipeline
  app.kubernetes.io/name: myapp-mysqld-exporter
  app.kubernetes.io/owner: myowner
  app.kubernetes.io/part-of: myapp
  k8s-app: myapp
  name: myapp-mysqld-exporter

namePrefix: myapp-

namespace: mynamespace

images:
- name: prom/mysqld-exporter
  newTag: v0.11.0
```

And finally this gets inherited by the environment-specific overlay (which is where the patches live):

```yaml
---
kind: Kustomization
apiVersion: kustomize.config.k8s.io/v1beta1

commonLabels:
  app.kubernetes.io/instance: review-myapp-mysqld-exporter
  app.mintel.com/pipeline-stage: review
  app.mintel.com/env: dev

namePrefix: review-

bases:
- ../bases

patches:
- patches/add-vars-to-exporter-main.yaml
- patches/remove-cloud-sql-proxy.yaml

configMapGenerator:
- env: properties/cluster.properties
  name: cluster-properties
  namespace: mynamespace
```

**Comment by jbrette** [comment]
I attempted to recreate your environment here: [issue_1343](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1343)

I did not see any special issue with namespace handling, aka you have to set the name and namespaces properly in the patches for it to work.

But I found **a real issue related in the order of the patches**. If the remove-cloud-sql-proxy.yaml is loaded first it works, in the other order it does not.
The root cause is [here](https://github.com/keleustes/kustomize/blob/allinone/k8sdeps/transformer/patch/conflictdetector.go#L174) when we are using the SMP conflict detector (aka when we are not patching CustomResourceDefinition).


```bash
kustomize build examples/issues/issue_1343/overlays/environment-specific/
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/instance: review-myapp-mysqld-exporter
    app.mintel.com/env: dev
    app.mintel.com/pipeline-stage: review
  name: review-cluster-properties-c928988k6m
  namespace: mynamespace
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/component: prometheus-exporter
    app.kubernetes.io/instance: review-myapp-mysqld-exporter
    app.kubernetes.io/managed-by: pipeline
    app.kubernetes.io/name: myapp-mysqld-exporter
    app.kubernetes.io/owner: myowner
    app.kubernetes.io/part-of: myapp
    app.mintel.com/env: dev
    app.mintel.com/pipeline-stage: review
    k8s-app: myapp
    name: myapp-mysqld-exporter
  name: review-myapp-mysqld-exporter
  namespace: mynamespace
spec:
  selector:
    matchLabels:
      app.kubernetes.io/component: prometheus-exporter
      app.kubernetes.io/instance: review-myapp-mysqld-exporter
      app.kubernetes.io/managed-by: pipeline
      app.kubernetes.io/name: myapp-mysqld-exporter
      app.kubernetes.io/owner: myowner
      app.kubernetes.io/part-of: myapp
      app.mintel.com/env: dev
      app.mintel.com/pipeline-stage: review
      k8s-app: myapp
      name: myapp-mysqld-exporter
  template:
    metadata:
      labels:
        app.kubernetes.io/component: prometheus-exporter
        app.kubernetes.io/instance: review-myapp-mysqld-exporter
        app.kubernetes.io/managed-by: pipeline
        app.kubernetes.io/name: myapp-mysqld-exporter
        app.kubernetes.io/owner: myowner
        app.kubernetes.io/part-of: myapp
        app.mintel.com/env: dev
        app.mintel.com/pipeline-stage: review
        k8s-app: myapp
        name: myapp-mysqld-exporter
    spec:
      containers:
      - args:
        - --web.listen-address=:9104
        - --log.level=info
        - --log.format=logger:stderr?json=true
        command:
        - /bin/mysqld_exporter
        env:
        - name: DATA_SOURCE_NAME
          value: user:password@(host:3306)/
        image: prom/mysqld-exporter:v0.11.0
        livenessProbe:
          httpGet:
            path: /
            port: 9104
          initialDelaySeconds: 30
        name: mysqld-exporter
        ports:
        - containerPort: 9104
          name: metrics
        readinessProbe:
          httpGet:
            path: /
            port: 9104
        resources:
          limits:
            cpu: 100m
            memory: 200Mi
          requests:
            cpu: 30m
            memory: 50Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 65534
```

**Comment by richardmarshall** [comment]
Like Jerome I was unable to reproduce the namespace requirement problem you're hitting. What version of kustomize are you running?

**Comment by jbrette** [comment]
@richardmarshall @bcbrockway I was able to reproduce the patch: $delete issue with kustomize 2.0.3, 2.1.0 and 3.0.2. I created a [PR](https://github.com/kubernetes-sigs/kustomize/pull/1356) to track it.

For the time being ensure that your patch: $delete is first in the list. What is also important to notice, is that if your project want to use different deployment (for instance argo-rollout), also the syntax would be the same, because you would be using a CRD, the "patch. $delete" would properly not work since it would be using JsonMergePatch and not StrategicMergePatch
